viadot/sources/sap_rfc.py

```diff
@@ -519,3 +651,513 @@ def to_df(self):
+        data will be downloaded in chunks.
+
+        Source: https://success.jitterbit.com/display/DOC/Guide+to+Using+RFC_READ_TABLE+to+Query+SAP+Tables#GuidetoUsingRFC_READ_TABLEtoQuerySAPTables-create-the-operation
+        - WHERE clause: 75 character limit
+        - SELECT: 512 character row limit
+
+        Returns:
+            pd.DataFrame: A DataFrame representing the result of the query provided in `PyRFC.query()`.
+        """
+        params = self._query
+        columns = self.select_columns_aliased
+        sep = self._query.get("DELIMITER")
+        fields_lists = self._query.get("FIELDS")
+        if len(fields_lists) > 1:
+            logger.info(f"Data will be downloaded in {len(fields_lists)} chunks.")
+        func = self.func
+        if sep is None:
+            # automatically find a working separator
+            SEPARATORS = [
+                "|",
+                "/t",
+                "#",
+                ";",
+                "@",
+                "%",
+                "^",
+                "`",
+                "~",
+                "{",
+                "}",
+                "$",
+            ]
+        else:
+            SEPARATORS = [sep]
+
+        for sep in SEPARATORS:
+            logger.info(f"Checking if separator '{sep}' works.")
+            if isinstance(self.rfc_unique_id[0], str):
+                # columns only for the first chunk and we add the rest later to avoid name conflicts
+                df = pd.DataFrame(columns=fields_lists[0])
+            else:
+                df = pd.DataFrame()
+            self._query["DELIMITER"] = sep
+            chunk = 1
+            row_index = 0
+            for fields in fields_lists:
+                logger.info(f"Downloading {chunk} data chunk...")
+                self._query["FIELDS"] = fields
+                try:
+                    response = self.call(func, **params)
+                except ABAPApplicationError as e:
+                    if e.key == "DATA_BUFFER_EXCEEDED":
+                        raise DataBufferExceeded(
+                            "Character limit per row exceeded. Please select fewer columns."
+                        )
+                    else:
+                        raise e
+                record_key = "WA"
+                data_raw = np.array(response["DATA"])
+
+                # if the reference columns are provided not necessary to remove any extra row.
+                if not isinstance(self.rfc_unique_id[0], str):
+                    row_index, data_raw, start = detect_extra_rows(
+                        row_index, data_raw, chunk, fields
+                    )
+                else:
+                    start = False
+
+                data_raw = catch_extra_separators(
+                    data_raw, record_key, sep, fields, self.replacement
+                )
+
+                records = np.array([row[record_key].split(sep) for row in data_raw])
```

Comment:
Seems like this could be refactor to iterate through the data only once and apply `catch_extra_separators()` row-wise (no need to use `numpy` then as well).